chain/client/src/stateless_validation/state_witness_producer.rs

```diff
@@ -163,88 +163,89 @@ impl Client {
-                    base_state,
-                    ..
-                }) => {
-                    implicit_transitions.push(ChunkStateTransition {
-                        block_hash: *block_hash,
-                        base_state,
-                        post_state_root: *self
-                            .chain
-                            .get_chunk_extra(block_hash, &shard_uid)?
-                            .state_root(),
-                    });
-                }
+
+        loop {
+            let header = self.chain.get_block_header(&current_block_hash)?;
+            if header.height() < prev_chunk_height_included {
+                return Err(Error::InvalidBlockHeight(prev_chunk_height_included));
+            }
+
+            let current_epoch_id = *header.epoch_id();
+            let current_shard_id =
+                self.epoch_manager.get_prev_shard_id(&current_block_hash, next_shard_id)?.0;
+            if current_shard_id != next_shard_id {
+                // If shard id changes, we need to get implicit state
+                // transition from current shard id to the next shard id.
+                let (chunk_state_transition, _, _) =
+                    self.get_state_transition(&current_block_hash, &next_epoch_id, next_shard_id)?;
+                implicit_transitions.push(chunk_state_transition);
+            }
+            next_epoch_id = current_epoch_id;
+            next_shard_id = current_shard_id;
+
+            if header.height() == prev_chunk_height_included {
+                break;
             }
+
+            // Add implicit state transition.
+            let (chunk_state_transition, _, _) = self.get_state_transition(
+                &current_block_hash,
+                &current_epoch_id,
+                current_shard_id,
+            )?;
+            implicit_transitions.push(chunk_state_transition);
+
+            current_block_hash = *header.prev_hash();
         }
 
+        let main_block = current_block_hash;
+        let epoch_id = next_epoch_id;
+        let shard_id = next_shard_id;
+        implicit_transitions.reverse();
+
+        // Get the main state transition.
+        let (main_transition, receipts_hash, contract_accesses) = if prev_chunk_header.is_genesis()
+        {
+            let shard_uid = self.epoch_manager.shard_id_to_uid(shard_id, &epoch_id)?;
+            (
+                ChunkStateTransition {
+                    block_hash: main_block,
+                    base_state: Default::default(),
+                    post_state_root: *self
+                        .chain
+                        .get_chunk_extra(&main_block, &shard_uid)?
+                        .state_root(),
+                },
+                hash(&borsh::to_vec::<[Receipt]>(&[]).unwrap()),
+                vec![],
+            )
+        } else {
+            self.get_state_transition(&main_block, &epoch_id, shard_id)?
+        };
```

Comment:
nit: It's a bit bulky, can you move it to a helper method? Either as a whole or just the genesis part.